File: 3. Behavioral Design Patterns/Command/Go/main.go
package main

import "fmt"

// Command Pattern Example in Go
// Remote executes commands on Light and Stereo

type Command interface{ Execute(); Undo() }

type Light struct{}
func (Light) On(){ fmt.Println("Light ON") }
func (Light) Off(){ fmt.Println("Light OFF") }

type Stereo struct{}
func (Stereo) On(){ fmt.Println("Stereo ON") }
func (Stereo) Off(){ fmt.Println("Stereo OFF") }

type LightOnCommand struct{ l Light }
func (c LightOnCommand) Execute(){ c.l.On() }
func (c LightOnCommand) Undo(){ c.l.Off() }

type LightOffCommand struct{ l Light }
func (c LightOffCommand) Execute(){ c.l.Off() }
func (c LightOffCommand) Undo(){ c.l.On() }

type StereoOnCommand struct{ s Stereo }
func (c StereoOnCommand) Execute(){ c.s.On() }
func (c StereoOnCommand) Undo(){ c.s.Off() }

type StereoOffCommand struct{ s Stereo }
func (c StereoOffCommand) Execute(){ c.s.Off() }
func (c StereoOffCommand) Undo(){ c.s.On() }

type RemoteButton struct{ cmd Command }
func (r *RemoteButton) Set(c Command){ r.cmd = c }
func (r RemoteButton) Press(){ r.cmd.Execute() }
func (r RemoteButton) PressUndo(){ r.cmd.Undo() }

func main(){
	l := Light{}; s := Stereo{}
	btn := RemoteButton{ cmd: LightOnCommand{l} }
	btn.Press(); btn.PressUndo()
	btn.Set(StereoOnCommand{s})
	btn.Press(); btn.PressUndo()
}

File: 3. Behavioral Design Patterns/Command/Rust/main.rs
// Command Pattern Example in Rust
// Remote executes commands on Light and Stereo

trait Command {
    fn execute(&mut self);
    fn undo(&mut self);
}

struct Light;
impl Light { fn on(&self){ println!("Light ON"); } fn off(&self){ println!("Light OFF"); } }

struct Stereo;
impl Stereo { fn on(&self){ println!("Stereo ON"); } fn off(&self){ println!("Stereo OFF"); } }

struct LightOnCommand { l: Light }
impl Command for LightOnCommand {
    fn execute(&mut self){ self.l.on(); }
    fn undo(&mut self){ self.l.off(); }
}

struct LightOffCommand { l: Light }
impl Command for LightOffCommand {
    fn execute(&mut self){ self.l.off(); }
    fn undo(&mut self){ self.l.on(); }
}

struct StereoOnCommand { s: Stereo }
impl Command for StereoOnCommand {
    fn execute(&mut self){ self.s.on(); }
    fn undo(&mut self){ self.s.off(); }
}

struct StereoOffCommand { s: Stereo }
impl Command for StereoOffCommand {
    fn execute(&mut self){ self.s.off(); }
    fn undo(&mut self){ self.s.on(); }
}

struct RemoteButton<C: Command> { cmd: C }
impl<C: Command> RemoteButton<C> {
    fn new(cmd: C) -> Self { Self { cmd } }
    fn press(&mut self){ self.cmd.execute(); }
    fn press_undo(&mut self){ self.cmd.undo(); }
}

fn main(){
    let light = Light; let stereo = Stereo;
    let mut btn = RemoteButton::new(LightOnCommand { l: light });
    btn.press(); btn.press_undo();
    let mut btn2 = RemoteButton::new(StereoOnCommand { s: stereo });
    btn2.press(); btn2.press_undo();
}
